File: sdks/typescript/src/session_runner.ts
/** paid-session/v1 broker control driver with idempotent HTTP refills. */

import { WebSocket } from "ws";

import { OpenClearinghouseClient, callerProof, parseWei, type SessionHandle } from "./client.js";
import { BrokerProtocolError, OpenClearinghouseError } from "./errors.js";

export interface SessionBalance {
  status: "ok" | "low" | "exhausted";
  claimed_units: number;
  debited_units: number;
  unit: string;
  runway_units: number;
  runway_seconds_estimate: number | null;
  will_refuse_next_refill: boolean;
}

export interface BrokerSession {
  sessionId: string;
  workId: string;
  state: string;
  runtime: {
    schema: string;
    public: Record<string, unknown>;
    grants: Record<string, unknown>[];
  };
  credential: string;
  leaseExpiresAt: string;
  balance: SessionBalance;
  control: {
    statusUrl: string;
    topupUrl: string;
    endUrl: string;
    eventsWs: string | null;
  };
}

export interface RefillEvent {
  refillSeq: number | null;
  expectedValueWei: bigint | null;
  fundedValueWei: bigint | null;
  capStatus: Record<string, unknown> | null;
  error: OpenClearinghouseError | null;
}

export interface WinddownEvent {
  reason: string;
  projectedEndAt: string | null;
}

export type RefillCallback = (event: RefillEvent) => void | Promise<void>;
export type WinddownCallback = (event: WinddownEvent) => void | Promise<void>;
export type CapExtensionCallback = (
  balance: SessionBalance,
) => number | null | Promise<number | null>;

export interface SessionRunnerOptions {
  client: OpenClearinghouseClient;
  handle: SessionHandle;
  onRefillSucceeded?: RefillCallback;
  onRefillRefused?: RefillCallback;
  onWinddownWarning?: WinddownCallback;
  approveCapExtension?: CapExtensionCallback;
  WebSocketCtor?: typeof WebSocket;
  fetch?: typeof fetch;
}

/** LOC's `POST /v1/sessions/{id}/close` response; wei fields are integer strings. */
interface FinalSettle {
  outcome: string;
  billed_value_wei: string;
  refund_wei: string;
}

export class SessionRunner {
  private readonly client: OpenClearinghouseClient;
  private handle: SessionHandle;
  private readonly onRefillSucceeded: RefillCallback | undefined;
  private readonly onRefillRefused: RefillCallback | undefined;
  private readonly onWinddownWarning: WinddownCallback | undefined;
  private readonly approveCapExtension: CapExtensionCallback | undefined;
  private readonly WS: typeof WebSocket;
  private readonly fetchImpl: typeof fetch;
  private ws: WebSocket | null = null;
  private broker: BrokerSession | null = null;
  private pendingRefillKey: string | null = null;
  private pendingRefill: Record<string, unknown> | null = null;
  private pendingRefillProof: string | null = null;
  private pendingRefillMaxTotalUnits: number | null = null;
  private finalSettle: FinalSettle | null = null;
  private finalWei: { billedValueWei: bigint; refundWei: bigint } | null = null;
  private closedResolve: (() => void) | null = null;
  private readonly closedPromise: Promise<void>;

  constructor(opts: SessionRunnerOptions) {
    this.client = opts.client;
    this.handle = opts.handle;
    this.onRefillSucceeded = opts.onRefillSucceeded;
    this.onRefillRefused = opts.onRefillRefused;
    this.onWinddownWarning = opts.onWinddownWarning;
    this.approveCapExtension = opts.approveCapExtension;
    this.WS = opts.WebSocketCtor ?? WebSocket;
    this.fetchImpl = opts.fetch ?? fetch;
    this.closedPromise = new Promise((resolve) => {
      this.closedResolve = resolve;
    });
  }

  get brokerSession(): BrokerSession | null {
    return this.broker;
  }

  get outcome(): string | null {
    return this.finalSettle?.outcome ?? null;
  }

  get billedValueWei(): bigint | null {
    return this.finalWei?.billedValueWei ?? null;
  }

  get refundWei(): bigint | null {
    return this.finalWei?.refundWei ?? null;
  }

  async start(): Promise<BrokerSession> {
    if (this.broker !== null) return this.broker;
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "Livepeer-Protocol": this.handle.protocol,
      "Livepeer-Capability": this.handle.capability,
      "Livepeer-Offering": this.handle.offering,
      "Livepeer-Request-Id": this.handle.requestId,
    };
    headers["Livepeer-Authorization"] = this.handle.spendAuthorization;
    headers["Livepeer-Caller-Proof"] = this.handle.callerProof;
    const response = await this.fetchImpl(
      `${this.handle.brokerUrl.replace(/\/+$/, "")}/v1/session`,
      {
        method: "POST",
        headers,
        body: this.handle.sessionOpenBody,
      },
    );
    if (!response.ok) throw protocolError(`broker session-open failed: ${String(response.status)}`);
    this.broker = parseOpen(await response.json(), this.handle);
    if (this.broker.control.eventsWs !== null) {
      this.ws = new this.WS(this.broker.control.eventsWs, {
        headers: { Authorization: `Bearer ${this.broker.credential}` },
      });
      this.ws.on("message", (data: Buffer | ArrayBuffer | Buffer[]) => {
        void this.handleWsMessage(data);
      });
    }
    return this.broker;
  }

  async status(): Promise<Record<string, unknown>> {
    const session = await this.start();
    const response = await this.fetchImpl(session.control.statusUrl, {
      headers: { Authorization: `Bearer ${session.credential}` },
    });
    if (!response.ok) throw protocolError(`broker status failed: ${String(response.status)}`);
    return (await response.json()) as Record<string, unknown>;
  }

  async onBalance(value: SessionBalance): Promise<void> {
    const balance = parseBalance(value);
    if (balance.will_refuse_next_refill) {
      await this.onWinddownWarning?.({
        reason: "broker_will_refuse_next_refill",
        projectedEndAt: null,
      });
      return;
    }
    if (balance.status !== "low") return;
    if (this.handle.session.refill === "bounded") {
      await this.onWinddownWarning?.({
        reason: "bounded_runway_exhausting",
        projectedEndAt: null,
      });
      return;
    }
    if (this.pendingRefill !== null) {
      if (this.pendingRefillMaxTotalUnits === null) {
        throw protocolError("wholesale refill lost its cumulative cap");
      }
      await this.refill(balance.claimed_units, this.pendingRefillMaxTotalUnits);
      return;
    }
    if (this.approveCapExtension === undefined) {
      await this.onWinddownWarning?.({
        reason: "wholesale_cap_extension_required",
        projectedEndAt: null,
      });
      return;
    }
    const newMax = await this.approveCapExtension(balance);
    if (newMax === null) {
      await this.onWinddownWarning?.({
        reason: "wholesale_cap_extension_declined",
        projectedEndAt: null,
      });
      return;
    }
    if (newMax <= this.handle.maxTotalUnits) {
      throw protocolError("wholesale cap extension must increase maxTotalUnits");
    }
    await this.refill(balance.claimed_units, newMax);
  }

  private async refill(observedUnits: number, maxTotalUnits: number): Promise<void> {
    const session = await this.start();
    this.pendingRefillKey ??= crypto.randomUUID();
    if (this.pendingRefill === null) {
      try {
        const refillOptions: {
          observedConsumedUnits: number;
          requestId: string;
          maxTotalUnits: number;
          workloadRequestDigest: string;
        } = {
          observedConsumedUnits: observedUnits,
          requestId: this.pendingRefillKey,
          maxTotalUnits,
          workloadRequestDigest: await sha256EmptyObject(),
        };
        this.pendingRefill = (await this.client.refillSession(
          this.handle.sessionId,
          refillOptions,
        )) as Record<string, unknown>;
        this.pendingRefillMaxTotalUnits = maxTotalUnits;
      } catch (error) {
        if (error instanceof OpenClearinghouseError) {
          await this.onRefillRefused?.({
            refillSeq: null,
            expectedValueWei: null,
            fundedValueWei: null,
            capStatus: null,
            error,
          });
          return;
        }
        throw error;
      }
    }
    const refill = this.pendingRefill;
    const response = await this.postTopup(session, refill);
    if (!response.ok) throw protocolError(`broker topup failed: ${String(response.status)}`);
    const acceptedRefill = refill;
    const brokerResult = (await response.json()) as { balance?: SessionBalance };
    if (this.pendingRefillProof === null) {
      throw protocolError("wholesale refill lost its caller proof");
    }
    this.handle = {
      ...this.handle,
      spendAuthorization: String(acceptedRefill.spend_authorization),
      callerProof: this.pendingRefillProof,
      maxTotalUnits,
    };
    if (brokerResult.balance?.will_refuse_next_refill) {
      await this.onWinddownWarning?.({
        reason: "broker_will_refuse_next_refill",
        projectedEndAt: null,
      });
    }
    await this.onRefillSucceeded?.({
      refillSeq: numberOrNull(acceptedRefill.refill_seq),
      expectedValueWei: weiOrNull(acceptedRefill.expected_value_wei),
      fundedValueWei: weiOrNull(acceptedRefill.funded_value_wei),
      capStatus: (acceptedRefill.cap_status as Record<string, unknown> | undefined) ?? null,
      error: null,
    });
    this.pendingRefill = null;
    this.pendingRefillKey = null;
    this.pendingRefillProof = null;
    this.pendingRefillMaxTotalUnits = null;
  }

  private async postTopup(
    session: BrokerSession,
    refill: Record<string, unknown>,
  ): Promise<Response> {
    const headers: Record<string, string> = {
      Authorization: `Bearer ${session.credential}`,
      "Content-Type": "application/json",
      "Livepeer-Request-Id": String(refill.request_id),
    };
    if (refill.accounting_mode !== "wholesale_account") {
      throw protocolError("LOC returned a non-wholesale refill");
    }
    if (typeof refill.spend_authorization !== "string" || refill.spend_authorization.length === 0) {
      throw protocolError("wholesale refill is missing spend authorization");
    }
    const authorization = refill.spend_authorization;
    const proof = await callerProof(authorization, this.handle.signCallerProof);
    if (proof === null) throw protocolError("wholesale refill is missing caller proof");
    headers["Livepeer-Authorization"] = authorization;
    headers["Livepeer-Caller-Proof"] = proof;
    this.pendingRefillProof = proof;
    return this.fetchImpl(session.control.topupUrl, {
      method: "POST",
      headers,
      body: "{}",
    });
  }

  private async handleWsMessage(data: Buffer | ArrayBuffer | Buffer[]): Promise<void> {
    const text = Buffer.isBuffer(data)
      ? data.toString("utf8")
      : Array.isArray(data)
        ? Buffer.concat(data).toString("utf8")
        : null;
    if (text === null) return;
    try {
      const payload = JSON.parse(text) as { type?: string; balance?: SessionBalance };
      if (payload.type === "session.balance" && payload.balance) {
        await this.onBalance(payload.balance);
      }
    } catch {
      // Ignore non-control frames.
    }
  }

  async close(args: { actualUnits: number; outcome?: string }): Promise<FinalSettle> {
    if (this.finalSettle !== null) return this.finalSettle;
    const session = await this.start();
    const response = await this.fetchImpl(session.control.endUrl, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${session.credential}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ reason: "gateway_close" }),
    });
    if (!response.ok) throw protocolError(`broker end failed: ${String(response.status)}`);
    const encodedSettlement = response.headers.get("livepeer-settlement");
    if (!encodedSettlement) {
      throw protocolError("broker end response missing Livepeer-Settlement");
    }
    let settlement: unknown;
    try {
      const bytes = Uint8Array.from(atob(encodedSettlement), (char) => char.charCodeAt(0));
      settlement = JSON.parse(new TextDecoder().decode(bytes)) as unknown;
    } catch {
      throw protocolError("broker end response has malformed Livepeer-Settlement");
    }
    this.ws?.close();
    const finalSettle = (await this.client.closeSession(this.handle.sessionId, {
      ...args,
      settlement,
    })) as FinalSettle;
    // Validate the wei fields eagerly so a malformed close surfaces here
    // rather than on a later getter access.
    this.finalWei = {
      billedValueWei: parseWei(finalSettle.billed_value_wei),
      refundWei: parseWei(finalSettle.refund_wei),
    };
    this.finalSettle = finalSettle;
    this.closedResolve?.();
    return this.finalSettle;
  }

  waitClosed(): Promise<void> {
    return this.closedPromise;
  }
}

async function sha256EmptyObject(): Promise<string> {
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode("{}"));
  return Array.from(new Uint8Array(digest), (value) => value.toString(16).padStart(2, "0")).join(
    "",
  );
}

function parseOpen(value: unknown, handle: SessionHandle): BrokerSession {
  const data = asRecord(value, "malformed broker open");
  const runtime = asRecord(data.runtime, "malformed broker runtime");
  const lease = asRecord(data.lease, "malformed broker lease");
  const control = asRecord(data.control, "malformed broker control");
  if (data.work_id !== handle.workId) throw protocolError("broker work_id mismatch");
  if (runtime.schema !== handle.session.descriptor_schema) {
    throw protocolError("broker descriptor schema mismatch");
  }
  const runtimePublic = asRecord(runtime.public, "malformed broker runtime.public");
  if (!Array.isArray(runtime.grants)) throw protocolError("malformed broker runtime.grants");
  const grants = runtime.grants.map((grant) => asRecord(grant, "malformed broker grant"));
  return {
    sessionId: requiredString(data, "session_id"),
    workId: requiredString(data, "work_id"),
    state: requiredString(data, "state"),
    runtime: {
      schema: requiredString(runtime, "schema"),
      public: runtimePublic,
      grants,
    },
    credential: requiredString(data, "credential"),
    leaseExpiresAt: requiredString(lease, "expires_at"),
    balance: parseBalance(data.balance),
    control: {
      statusUrl: requiredString(control, "status_url"),
      topupUrl: requiredString(control, "topup_url"),
      endUrl: requiredString(control, "end_url"),
      eventsWs: optionalString(control, "events_ws"),
    },
  };
}

function parseBalance(value: unknown): SessionBalance {
  const balance = asRecord(value, "malformed broker balance");
  if (!["ok", "low", "exhausted"].includes(String(balance.status))) {
    throw protocolError("malformed broker balance");
  }
  return balance as unknown as SessionBalance;
}

function requiredString(value: Record<string, unknown>, key: string): string {
  const result = value[key];
  if (typeof result !== "string" || result.length === 0) throw protocolError(`missing ${key}`);
  return result;
}

function optionalString(value: Record<string, unknown>, key: string): string | null {
  const result = value[key];
  if (result == null) return null;
  if (typeof result !== "string" || result.length === 0) throw protocolError(`invalid ${key}`);
  return result;
}

function asRecord(value: unknown, message: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw protocolError(message);
  }
  return value as Record<string, unknown>;
}

function protocolError(message: string): BrokerProtocolError {
  return new BrokerProtocolError(message);
}

function numberOrNull(value: unknown): number | null {
  return value == null ? null : Number(value);
}

function weiOrNull(value: unknown): bigint | null {
  return value == null ? null : parseWei(value as string | number);
}

File: sdks/rust/tests/session_runner.rs
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;

use base64::Engine as _;
use livepeer_open_clearinghouse_sdk::{
    Client, ClientOptions, SessionAxes, SessionBalance, SessionHandle, SessionRunner,
    SessionRunnerOptions,
};
use serde_json::json;
use tokio::sync::Mutex;
use wiremock::matchers::{body_json, body_partial_json, header, method, path};
use wiremock::{Mock, MockServer, ResponseTemplate};

const SID: &str = "11111111-1111-1111-1111-111111111111";

fn handle(broker: &MockServer, refill: &str) -> SessionHandle {
    SessionHandle {
        session_id: SID.to_string(),
        request_id: "open-request".to_string(),
        work_id: "wid".to_string(),
        broker_url: broker.uri(),
        protocol: "paid-session/v1".to_string(),
        capability: "livepeer:test".to_string(),
        offering: "default".to_string(),
        session: SessionAxes {
            descriptor_schema: "livepeer-session-test/v1".to_string(),
            attachment: "external".to_string(),
            metering: "runner-reported".to_string(),
            refill: refill.to_string(),
        },
        session_params: json!({"room": "alpha"}),
        spend_authorization: Some(base64::engine::general_purpose::STANDARD.encode("initial")),
        accounting_mode: "wholesale_account".to_string(),
        caller_proof: Some("INITIAL-PROOF".to_string()),
        session_open_body: Vec::new(),
        max_total_units: 100,
        expected_value_wei: 100_000,
        funded_value_wei: 100_000,
        refill_endpoint: format!("/v1/sessions/{SID}/refill"),
        close_endpoint: format!("/v1/sessions/{SID}/close"),
        opened_at: "2026-08-20T00:00:00Z".to_string(),
    }
}

fn balance(status: &str, refuse: bool) -> serde_json::Value {
    json!({
        "status": status, "claimed_units": 80u64, "debited_units": 80u64,
        "unit": "participant_minutes", "runway_units": 20u64,
        "runway_seconds_estimate": 1200u64, "will_refuse_next_refill": refuse
    })
}

async fn mount_open(broker: &MockServer, schema: &str) {
    Mock::given(method("POST"))
        .and(path("/v1/session"))
        .and(header("Livepeer-Protocol", "paid-session/v1"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "session_id": "broker-session", "work_id": "wid", "state": "active",
            "runtime": {"schema": schema, "public": {}, "grants": []},
            "credential": "credential", "lease": {"expires_at": "2026-08-21T00:00:00Z"},
            "balance": balance("ok", false),
            "control": {
                "status_url": format!("{}/status", broker.uri()),
                "topup_url": format!("{}/topup", broker.uri()),
                "end_url": format!("{}/end", broker.uri())
            }
        })))
        .mount(broker)
        .await;
}

#[tokio::test]
async fn paid_session_v1_open_refill_and_close() {
    let broker = MockServer::start().await;
    mount_open(&broker, "livepeer-session-test/v1").await;
    let topup_attempts = Arc::new(AtomicUsize::new(0));
    let attempts = topup_attempts.clone();
    Mock::given(method("POST"))
        .and(path("/topup"))
        .and(header("Authorization", "Bearer credential"))
        .and(header("Livepeer-Request-Id", "refill-request"))
        .respond_with(move |_: &wiremock::Request| {
            if attempts.fetch_add(1, Ordering::SeqCst) == 0 {
                ResponseTemplate::new(503)
            } else {
                ResponseTemplate::new(200).set_body_json(json!({"balance": balance("ok", false)}))
            }
        })
        .expect(2)
        .mount(&broker)
        .await;
    Mock::given(method("GET"))
        .and(path("/status"))
        .and(header("Authorization", "Bearer credential"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({"state": "active"})))
        .expect(1)
        .mount(&broker)
        .await;
    Mock::given(method("POST"))
        .and(path("/end"))
        .respond_with(ResponseTemplate::new(204).insert_header(
            "Livepeer-Settlement",
            "eyJwYXlsb2FkIjp7fSwic2lnbmF0dXJlIjp7fX0=",
        ))
        .expect(1)
        .mount(&broker)
        .await;

    let loc = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path(format!("/v1/sessions/{SID}/refill")))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "work_id": "loc-auth:revision",
            "request_id": "refill-request", "refill_seq": 1u64,
            "spend_authorization": "cmV2aXNpb24=", "accounting_mode": "wholesale_account",
            "expected_value_wei": "50000",
            "funded_value_wei": "50000", "cap_status": null
        })))
        .expect(1)
        .mount(&loc)
        .await;
    Mock::given(method("POST"))
        .and(path(format!("/v1/sessions/{SID}/close")))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "outcome": "EXACT", "billed_value_wei": "150000", "refund_wei": "0"
        })))
        .mount(&loc)
        .await;

    let client = Client::new(
        ClientOptions::new(loc.uri(), "pymth_test")
            .with_caller_proof("public-key", Arc::new(|_| Ok("PROOF".to_string()))),
    )
    .unwrap();
    let first = SessionRunner::start(SessionRunnerOptions::new(
        client.clone(),
        handle(&broker, "extensible"),
    ))
    .await
    .unwrap();
    drop(first);
    let mut options = SessionRunnerOptions::new(client, handle(&broker, "extensible"));
    options.approve_cap_extension = Some(Arc::new(|_| Box::pin(async { Some(200) })));
    let runner = SessionRunner::start(options).await.unwrap();
    assert_eq!(runner.status().await.unwrap()["state"], "active");
    runner
        .on_balance(serde_json::from_value(balance("low", false)).unwrap())
        .await;
    runner
        .on_balance(serde_json::from_value(balance("low", false)).unwrap())
        .await;
    let outcome = runner.close(150).await.unwrap();
    assert_eq!(outcome.outcome, "EXACT");
    assert_eq!(topup_attempts.load(Ordering::SeqCst), 2);
}

#[tokio::test]
#[ignore = "loc-0m4.4 removes obsolete recipient-rotation ticket coverage"]
async fn recipient_rotation_uses_fresh_intent_and_declared_rebind() {
    let broker = MockServer::start().await;
    mount_open(&broker, "livepeer-session-test/v1").await;
    Mock::given(method("POST"))
        .and(path("/topup"))
        .and(header("Livepeer-Payment", "OLD"))
        .respond_with(
            ResponseTemplate::new(409).insert_header("Livepeer-Error", "recipient_rotated"),
        )
        .expect(1)
        .mount(&broker)
        .await;
    Mock::given(method("POST"))
        .and(path("/topup"))
        .and(header("Livepeer-Payment", "NEW"))
        .and(header("Livepeer-Request-Id", "replacement"))
        .and(header("Livepeer-Rebind-From", "wid"))
        .respond_with(
            ResponseTemplate::new(200).set_body_json(json!({"balance": balance("ok", false)})),
        )
        .expect(1)
        .mount(&broker)
        .await;

    let loc = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path(format!("/v1/sessions/{SID}/refill")))
        .and(body_partial_json(json!({"rebind_from": "wid"})))
        .and(body_partial_json(
            json!({"replaces_request_id": "rejected"}),
        ))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "work_id": "successor", "request_id": "replacement",
            "payment_envelope": "NEW", "rebind_from": "wid"
        })))
        .expect(1)
        .mount(&loc)
        .await;
    Mock::given(method("POST"))
        .and(path(format!("/v1/sessions/{SID}/refill")))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "work_id": "wid", "request_id": "rejected", "payment_envelope": "OLD"
        })))
        .expect(1)
        .mount(&loc)
        .await;

    let client = Client::new(ClientOptions::new(loc.uri(), "pymth_test")).unwrap();
    let warnings = Arc::new(Mutex::new(Vec::new()));
    let captured = warnings.clone();
    let mut options = SessionRunnerOptions::new(client, handle(&broker, "extensible"));
    options.on_winddown_warning = Some(Arc::new(move |event| {
        let captured = captured.clone();
        Box::pin(async move { captured.lock().await.push(event.reason) })
    }));
    let runner = SessionRunner::start(options).await.unwrap();
    runner
        .on_balance(serde_json::from_value(balance("low", false)).unwrap())
        .await;
    assert_eq!(runner.broker_session().await.work_id, "successor");
    assert!(warnings.lock().await.is_empty());
}

#[tokio::test]
#[ignore = "loc-0m4.4 removes obsolete recipient-rebind ticket coverage"]
async fn declared_rebind_refusal_drains_once() {
    let broker = MockServer::start().await;
    mount_open(&broker, "livepeer-session-test/v1").await;
    Mock::given(method("POST"))
        .and(path("/topup"))
        .and(header("Livepeer-Payment", "OLD"))
        .respond_with(
            ResponseTemplate::new(409).insert_header("Livepeer-Error", "recipient_rotated"),
        )
        .expect(1)
        .mount(&broker)
        .await;
    Mock::given(method("POST"))
        .and(path("/topup"))
        .and(header("Livepeer-Payment", "NEW"))
        .and(header("Livepeer-Rebind-From", "wid"))
        .respond_with(ResponseTemplate::new(409).insert_header("Livepeer-Error", "rebind_refused"))
        .expect(1)
        .mount(&broker)
        .await;

    let loc = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path(format!("/v1/sessions/{SID}/refill")))
        .and(body_partial_json(json!({"rebind_from": "wid"})))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "work_id": "successor", "request_id": "replacement",
            "payment_envelope": "NEW", "rebind_from": "wid"
        })))
        .expect(1)
        .mount(&loc)
        .await;
    Mock::given(method("POST"))
        .and(path(format!("/v1/sessions/{SID}/refill")))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "work_id": "wid", "request_id": "rejected", "payment_envelope": "OLD"
        })))
        .expect(1)
        .mount(&loc)
        .await;

    let warnings = Arc::new(Mutex::new(Vec::new()));
    let captured = warnings.clone();
    let client = Client::new(ClientOptions::new(loc.uri(), "pymth_test")).unwrap();
    let mut options = SessionRunnerOptions::new(client, handle(&broker, "extensible"));
    options.on_winddown_warning = Some(Arc::new(move |event| {
        let captured = captured.clone();
        Box::pin(async move { captured.lock().await.push(event.reason) })
    }));
    let runner = SessionRunner::start(options).await.unwrap();
    runner
        .on_balance(serde_json::from_value(balance("low", false)).unwrap())
        .await;
    assert_eq!(*warnings.lock().await, ["payment_unrecoverable"]);
}

#[tokio::test]
async fn bounded_and_refusal_warning_balances_drain() {
    let broker = MockServer::start().await;
    mount_open(&broker, "livepeer-session-test/v1").await;
    let loc = MockServer::start().await;
    let warnings = Arc::new(Mutex::new(Vec::new()));
    let captured = warnings.clone();
    let client = Client::new(ClientOptions::new(loc.uri(), "pymth_test")).unwrap();
    let mut options = SessionRunnerOptions::new(client, handle(&broker, "bounded"));
    options.on_winddown_warning = Some(Arc::new(move |event| {
        let captured = captured.clone();
        Box::pin(async move { captured.lock().await.push(event.reason) })
    }));
    let runner = SessionRunner::start(options).await.unwrap();
    runner
        .on_balance(serde_json::from_value::<SessionBalance>(balance("low", false)).unwrap())
        .await;
    runner
        .on_balance(serde_json::from_value::<SessionBalance>(balance("ok", true)).unwrap())
        .await;
    assert_eq!(
        *warnings.lock().await,
        [
            "bounded_runway_exhausting",
            "broker_will_refuse_next_refill"
        ]
    );
}

#[tokio::test]
async fn wholesale_low_balance_requires_explicit_cap_approval() {
    let broker = MockServer::start().await;
    mount_open(&broker, "livepeer-session-test/v1").await;
    let loc = MockServer::start().await;
    let warnings = Arc::new(Mutex::new(Vec::new()));
    let captured = warnings.clone();
    let mut wholesale = handle(&broker, "extensible");
    wholesale.spend_authorization = Some("aW5pdGlhbA==".to_string());
    wholesale.accounting_mode = "wholesale_account".to_string();
    wholesale.caller_proof = Some("INITIAL-PROOF".to_string());
    let client = Client::new(ClientOptions::new(loc.uri(), "pymth_test")).unwrap();
    let mut options = SessionRunnerOptions::new(client, wholesale);
    options.on_winddown_warning = Some(Arc::new(move |event| {
        let captured = captured.clone();
        Box::pin(async move { captured.lock().await.push(event.reason) })
    }));
    let runner = SessionRunner::start(options).await.unwrap();
    runner
        .on_balance(serde_json::from_value(balance("low", false)).unwrap())
        .await;
    assert_eq!(*warnings.lock().await, ["wholesale_cap_extension_required"]);
}

#[tokio::test]
async fn wholesale_cap_revision_commits_and_signs_exact_body() {
    let broker = MockServer::start().await;
    mount_open(&broker, "livepeer-session-test/v1").await;
    Mock::given(method("POST"))
        .and(path("/topup"))
        .and(header("Livepeer-Authorization", "c3VjY2Vzc29y"))
        .and(header("Livepeer-Caller-Proof", "SUCCESSOR-PROOF"))
        .and(body_json(json!({})))
        .respond_with(
            ResponseTemplate::new(200).set_body_json(json!({"balance": balance("ok", false)})),
        )
        .expect(1)
        .mount(&broker)
        .await;
    let loc = MockServer::start().await;
    Mock::given(method("POST"))
        .and(path(format!("/v1/sessions/{SID}/refill")))
        .and(body_partial_json(json!({
            "observed_consumed_units": 80u64,
            "max_total_units": 200u64,
            "workload_request_digest": "44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a"
        })))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "work_id": "loc-auth:revision", "request_id": "revision-request",
            "refill_seq": 1u64, "payment_envelope": null,
            "spend_authorization": "c3VjY2Vzc29y", "accounting_mode": "wholesale_account",
            "expected_value_wei": "0", "funded_value_wei": "0", "cap_status": null
        })))
        .expect(1)
        .mount(&loc)
        .await;
    let signer = Arc::new(|bytes: &[u8]| {
        assert_eq!(bytes, b"successor");
        Ok("SUCCESSOR-PROOF".to_string())
    });
    let client = Client::new(
        ClientOptions::new(loc.uri(), "pymth_test").with_caller_proof("public-key", signer),
    )
    .unwrap();
    let mut wholesale = handle(&broker, "extensible");
    wholesale.spend_authorization = Some("aW5pdGlhbA==".to_string());
    wholesale.accounting_mode = "wholesale_account".to_string();
    wholesale.caller_proof = Some("INITIAL-PROOF".to_string());
    let mut options = SessionRunnerOptions::new(client, wholesale);
    options.approve_cap_extension = Some(Arc::new(|_| Box::pin(async { Some(200) })));
    let runner = SessionRunner::start(options).await.unwrap();
    runner
        .on_balance(serde_json::from_value(balance("low", false)).unwrap())
        .await;
}

#[tokio::test]
async fn descriptor_mismatch_fails_closed() {
    let broker = MockServer::start().await;
    mount_open(&broker, "wrong/v1").await;
    let loc = MockServer::start().await;
    let client = Client::new(ClientOptions::new(loc.uri(), "pymth_test")).unwrap();
    let result = SessionRunner::start(SessionRunnerOptions::new(
        client,
        handle(&broker, "extensible"),
    ))
    .await;
    assert!(result.is_err());
}
